sdk/core/azure-core/inc/azure/core/http/http.hpp

```diff
@@ -135,154 +155,216 @@ namespace Azure { namespace Core { namespace Http {
-        std::map<std::string, std::string> left,
-        std::map<std::string, std::string> const& right)
-    {
-      left.insert(right.begin(), right.end());
-      return left;
-    }
-
-    std::string GetQueryString() const;
-
     // This value can be used to override the default value that an http transport adapter uses to
     // read and upload chunks of data from the payload body stream. If it is not set, the transport
     // adapter will decide chunk size.
     int64_t m_uploadChunkSize = 0;
 
+    void StartRetry(); // only called by retry policy
+    void StopRetry(); // only called by retry policy
+
   public:
-    explicit Request(
-        HttpMethod httpMethod,
-        std::string const& url,
-        BodyStream* bodyStream,
-        bool downloadViaStream)
-        : m_method(std::move(httpMethod)), m_url(url), m_bodyStream(bodyStream),
+    explicit Request(HttpMethod httpMethod, URL url, BodyStream* bodyStream, bool downloadViaStream)
+        : m_method(std::move(httpMethod)), m_url(std::move(url)), m_bodyStream(bodyStream),
           m_retryModeEnabled(false), m_isDownloadViaStream(downloadViaStream)
     {
     }
 
-    explicit Request(HttpMethod httpMethod, std::string const& url, BodyStream* bodyStream)
-        : Request(httpMethod, url, bodyStream, false)
+    explicit Request(HttpMethod httpMethod, URL url, BodyStream* bodyStream)
+        : Request(httpMethod, std::move(url), bodyStream, false)
     {
     }
 
     // Typically used for GET with no request body that can return bodyStream
-    explicit Request(HttpMethod httpMethod, std::string const& url, bool downloadViaStream)
-        : Request(httpMethod, url, NullBodyStream::GetNullBodyStream(), downloadViaStream)
+    explicit Request(HttpMethod httpMethod, URL url, bool downloadViaStream)
+        : Request(
+            httpMethod,
+            std::move(url),
+            NullBodyStream::GetNullBodyStream(),
+            downloadViaStream)
     {
     }
 
     // Typically used for GET with no request body.
-    explicit Request(HttpMethod httpMethod, std::string const& url)
-        : Request(httpMethod, url, NullBodyStream::GetNullBodyStream(), false)
+    explicit Request(HttpMethod httpMethod, URL url)
+        : Request(httpMethod, std::move(url), NullBodyStream::GetNullBodyStream(), false)
     {
     }
 
-    // Methods used to build HTTP request
-    void AppendPath(std::string const& path);
-    void AddQueryParameter(std::string const& name, std::string const& value);
     void AddHeader(std::string const& name, std::string const& value);
-    void StartRetry(); // only called by retry policy
     void SetUploadChunkSize(int64_t size) { this->m_uploadChunkSize = size; }
 
     // Methods used by transport layer (and logger) to send request
     HttpMethod GetMethod() const;
-    std::string GetEncodedUrl() const; // should call URL encode
-    std::string GetHost() const;
     std::map<std::string, std::string> GetHeaders() const;
```

Comment:
How about const reference?

Reply:
a headers/query param copy is returned because of the retry feature.
When retry is enabled, the values from two maps are merged to be returned and all values from retry overrides values from the normal map.

Reply:
okay